go-controller/pkg/testing/libovsdb/matchers.go

```diff
@@ -203,41 +203,60 @@ func testDataEqual(x, y TestData, ignoreUUIDs bool) bool {
 
 // HaveData matches expected libovsdb models with named UUIDs
 func HaveData(expected ...TestData) gomegatypes.GomegaMatcher {
-	return haveData(false, true, expected)
+	return haveData(false, true, expected, gomega.ConsistOf)
 }
 
 // HaveDataIgnoringUUIDs matches expected libovsdb models ignoring UUIDs
 func HaveDataIgnoringUUIDs(expected ...TestData) gomegatypes.GomegaMatcher {
-	return haveData(true, false, expected)
+	return haveData(true, false, expected, gomega.ConsistOf)
 }
 
 // HaveDataExact matches expected libovsdb models exactly
 func HaveDataExact(expected ...TestData) gomegatypes.GomegaMatcher {
-	return haveData(false, false, expected)
+	return haveData(false, false, expected, gomega.ConsistOf)
 }
 
 func HaveEmptyData() gomegatypes.GomegaMatcher {
 	return gomega.WithTransform(getTestDataFromClientCache, gomega.BeEmpty())
 }
 
-func haveData(ignoreUUIDs, nameUUIDs bool, expected []TestData) gomegatypes.GomegaMatcher {
+func haveData(ignoreUUIDs, nameUUIDs bool, expected []TestData, matcherFunction func(elements ...interface{}) gomegatypes.GomegaMatcher) gomegatypes.GomegaMatcher {
 	expected = flatten(expected)
 	matchers := testDataMatchers(ignoreUUIDs, expected)
 
-	transform := func(client libovsdbclient.Client) []TestData {
-		actual := getTestDataFromClientCache(client)
-		if nameUUIDs {
-			expectedCopy := copystructure.Must(copystructure.Copy(expected)).([]TestData)
-			actualCopy := copystructure.Must(copystructure.Copy(actual)).([]TestData)
-			matchAndReplaceNamedUUIDs(actualCopy, expectedCopy)
-			for i, m := range matchers {
-				m.expected = expectedCopy[i]
-			}
-			return actualCopy
+	if nameUUIDs {
+		namedUUIDTransformWrapper := func(client libovsdbclient.Client) []TestData {
+			return transformTestDataWithUUIDs(client, expected, matchers)
 		}
-		return actual
+		return gomega.WithTransform(namedUUIDTransformWrapper, matcherFunction(matchers))
+	}
+
+	return gomega.WithTransform(getTestDataFromClientCache, matcherFunction(matchers))
+}
+
+// ContainData succeeds if actual contains the passed in elements.
+// The ordering of the elements does not matter.
+func ContainData(expected ...TestData) gomegatypes.GomegaMatcher {
+	return haveData(false, true, expected, gomega.ContainElements)
+}
+
+// ContainDataIgnoringUUIDs succeeds if actual contains the passed in elements,
+// ignoring the elements UUIDs.
+// The ordering of the elements does not matter.
+func ContainDataIgnoringUUIDs(expected ...TestData) gomegatypes.GomegaMatcher {
+	return haveData(true, false, expected, gomega.ContainElements)
+}
```

Comment:
We have been, or at least I have been, avoiding this type of matcher. It leads to not verifying the full contents of the database in a way that in-adverted changes can happen to the database that should have been verified not to happen. Even if you are sure that would not happen in your test, other people can start using this matcher without thinking things through.
Please, I won't be strict about it but reconsider your approach. What prevents in your test to verify the full contents of the database? Verifying the full contents of the database leads to a stronger test.

Reply:
I understand your point. My motivation is to see a clear error when something goes wrong, instead of having to figure out what thing is missing/added in a huge haystack.

Would it make sense then to assert both ? First the "localized" matcher, which would give out the exact error, then the "general" matcher which would assert the full desired state of the DB ? It may seem redundant but I feel it does help out the developer.

Reply:
> I understand your point. My motivation is to see a clear error when something goes wrong, instead of having to figure out what thing is missing/added in a huge haystack.

Certainly, processing the output of failed matcher over a big database requires a bit of practice. It still has a `missing things in expected` and `missing things in actual` that you could focus on? We might be able to customize its output to just that?

> Would it make sense then to assert both ? First the "localized" matcher, which would give out the exact error, then the "general" matcher which would assert the full desired state of the DB ? It may seem redundant but I feel it does help out the developer.

Would there be a way to express that with a single overall matcher that combines both things in a way that `ContainData` could not be used individually? 

Or perhaps it would be less harmful to have a `DoesNotContainData` matcher directly.

Reply:
I feel the ginkgo way of doing this is not like that; that falls under the `Should` / `ShouldNot` / `To` / `NotTo` [assertion interface](https://github.com/ovn-org/ovn-kubernetes/blob/e19c20762c54f51615cd8ecff86eaea35e4b6efb/go-controller/vendor/github.com/onsi/gomega/gomega_dsl.go#L393).

Let me think this through a little bit, you gave me plenty to think about.

Reply:
I don't wanna get in your way, just making quick suggestions to see if they make any sense to you.

Reply:
I'm matching the original node balancer is not present **but** have also added matching against the final expected state of the cluster. 

If you think the focused matcher I've introduced is not required, I can remove the commit that has added it. 

If this approach seems preferable to you, I will also gladly squash the top-most commit.

Reply:
@jcaamano ^